wandb/sdk/wandb_run.py

```diff
@@ -1951,260 +1711,54 @@ def _on_finish(self) -> None:
         self.history._flush()
 
         self._console_stop()  # TODO: there's a race here with jupyter console logging
-        if not self._settings._silent:
-            as_html = self._settings._jupyter and ipython.in_jupyter()
-            if self._backend:
-                pid = self._backend._internal_pid
-                status_str = "Waiting for W&B process to finish, PID {}... ".format(pid)
-            if not self._exit_code:
-                status = "(success)."
-                if as_html:
-                    status = f'<strong style="color:green">{status}</strong>'
-                status_str += status
-            else:
-                status = "(failed {}).".format(self._exit_code)
-                if as_html:
-                    status = f'<strong style="color:red">{status}</strong>'
-                status_str += status
-                if not self._settings._offline:
-                    status_str += " Press ctrl-c to abort syncing."
-            if as_html:
-                sep = "<br/>" if not self._quiet else ""
-                ipython.display_html(sep + status_str)
-            else:
-                print("")
-                wandb.termlog(status_str)
 
-        # telemetry could have changed, publish final data
         if self._backend and self._backend.interface:
+            # telemetry could have changed, publish final data
             self._backend.interface._publish_telemetry(self._telemetry_obj)
 
-        # TODO: we need to handle catastrophic failure better
-        # some tests were timing out on sending exit for reasons not clear to me
-        if self._backend and self._backend.interface:
+            # TODO: we need to handle catastrophic failure better
+            # some tests were timing out on sending exit for reasons not clear to me
             self._backend.interface.publish_exit(self._exit_code)
 
-        # Wait for data to be synced
-        self._poll_exit_response = self._wait_for_finish()
+        print("")
+        self._footer_exit_status_info(
+            self._exit_code, settings=self._settings, printer=self._printer
+        )
 
-        if self._backend and self._backend.interface:
-            ret = self._backend.interface.communicate_get_summary()
-            if ret:
-                self._final_summary = proto_util.dict_from_proto_list(ret.item)
+        while not (self._poll_exit_response and self._poll_exit_response.done):
```

Comment:
could this ever become an infinite loop?

Reply:
in theory yes, but we have time-out